pkg/parameters/aws/aws.go

```diff
@@ -0,0 +1,213 @@
+package aws
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/aws/aws-sdk-go/service/ssm"
+	"gopkg.in/yaml.v2"
+	"k8s.io/klog"
+)
+
+const (
+	awsDetectionTimeout  = 1 * time.Second
+	awsTimeout           = 10 * time.Second
+	ssmMaxResults        = 10
+	ssmParameterBaseName = "config"
+)
+
+type AWSParameters struct {
+	base string
+	ec2  *ec2.EC2
+	ssm  *ssm.SSM
+}
+
+func detectRegion() string {
+	session, err := session.NewSession()
+	if err != nil {
+		klog.Warningf("creating session to autodetect AWS region: %v", err)
+		return ""
+	}
+	client := ec2metadata.New(session)
+
+	ctx, cancel := context.WithTimeout(aws.BackgroundContext(), awsDetectionTimeout)
+	defer cancel()
+	region, err := client.RegionWithContext(ctx)
+	if err != nil {
+		klog.Warningf("trying to autodetect AWS region: %v", err)
+		return ""
+	}
+
+	klog.V(2).Infof("detected AWS region: %q", region)
+	return region
+}
+
+func detectInstanceID() string {
+	session, err := session.NewSession()
+	if err != nil {
+		klog.Warningf("creating session to autodetect AWS region: %v", err)
+		return ""
+	}
+	client := ec2metadata.New(session)
+
+	ctx, cancel := context.WithTimeout(aws.BackgroundContext(), awsDetectionTimeout)
+	defer cancel()
+	instanceID, err := client.GetMetadataWithContext(ctx, "instance-id")
+	if err != nil {
+		klog.Warningf("trying to autodetect AWS instance ID: %v", err)
+		return ""
+	}
+
+	klog.V(2).Infof("detected AWS instance ID: %q", instanceID)
+	return instanceID
+}
+
+func getAWSConfig() *aws.Config {
+	region := os.Getenv("AWS_REGION")
+	if region == "" {
+		region = os.Getenv("AWS_DEFAULT_REGION")
+	}
+	if region == "" {
+		region = detectRegion()
```

Comment:
shouldn't we check somewhere if `detectRegion()` returned actual region, not `""`? Or is it fine to fail later?

Reply:
That makes sense, it'd fail later on, but it's better to catch this early. :+1: